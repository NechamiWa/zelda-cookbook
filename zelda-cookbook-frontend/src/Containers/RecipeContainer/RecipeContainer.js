import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import RecipeCard from '../../Components/RecipeCard/RecipeCard';
import DubiousFood from '../../assets/DubiousFood.png';
import './RecipeContainer.css';

export class RecipeContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: true
    };
  }

  // componentDidMount() {
  //   this.setState({ loading: false })
  // }

  filterByIngredient = recipes => {
    const { ingredientFilter } = this.props;

    if (ingredientFilter.length) {
      const filtered = recipes.filter(recipe => {
        let match = 0;
        ingredientFilter.forEach(ingredient => {
          for (let i = 1; i <= 5; i++) {
            if (parseInt(ingredient.id, 10) === recipe[`ingredient${i}_id`]) {
              match++;
            }
          }
        });
        return match === ingredientFilter.length;
      });
      return filtered;
    } else {
      return recipes;
    }
  };

  filterByName = recipes => {
    const { nameFilter } = this.props;
    if (nameFilter) {
      return recipes.filter(recipe =>
        recipe.name.toLowerCase().includes(nameFilter.toLowerCase())
      );
    } else {
      return recipes;
    }
  };

  filterByType = recipes => {
    const { typeFilter } = this.props;
    if (typeFilter) {
      return recipes.filter(recipe => recipe.type === typeFilter);
    } else {
      return recipes;
    }
  };

  filterRecipes = allRecipes => {
    const onceFiltered = this.filterByIngredient(allRecipes);
    const twiceFiltered = this.filterByName(onceFiltered);
    const thriceFiltered = this.filterByType(twiceFiltered);
    return thriceFiltered;
  };

  renderRecipes = () => {
    const recipes = this.filterRecipes(this.props.recipes);

    if (recipes.length) {
      return recipes.map(recipe => <RecipeCard key={recipe.id} {...recipe} />);
    } else {
      return (
        <div id="loading">
          <h1>No Recipe Found!<img src={ DubiousFood } alt="Dubious Food"></img></h1>
        </div>
      );
    }
  };

  render() {
    return <div id="all-recipe-container">{this.renderRecipes()}</div>;
  }
}

const { arrayOf, shape, number, string, array } = PropTypes;

RecipeContainer.propTypes = {
  recipes: arrayOf(
    shape({
      id: number,
      category: string,
      hearts: string,
      name: string,
      notes: string,
      resale: string,
      type: string,
      type_image: string,
      duration: string,
      strength: string,
      ingredient1: string,
      ingredient2: string,
      ingredient3: string,
      ingredient4: string,
      ingredient5: string,
      ingredient1_image: string,
      ingredient2_image: string,
      ingredient3_image: string,
      ingredient4_image: string,
      ingredient5_image: string,
      ingredient1_id: number,
      ingredient2_id: number,
      ingredient3_id: number,
      ingredient4_id: number,
      ingredient5_id: number,
      created_at: string,
      updated_at: string
    })
  ),
  ingredientFilter: array,
  nameFilter: string,
  typeFilter: string
};

export const MSTP = props => {
  const { recipes, ingredientFilter, nameFilter, typeFilter } = props;

  return {
    recipes,
    ingredientFilter,
    nameFilter,
    typeFilter
  };
};

export default connect(MSTP)(RecipeContainer);
