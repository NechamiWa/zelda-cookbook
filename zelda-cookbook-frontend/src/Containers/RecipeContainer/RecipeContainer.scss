$gif-inset: rgb(67,110,82);

#loading {
  background-image: url('https://68.media.tumblr.com/7a5f3925e8e8351619759a2a9566bb03/tumblr_omsjs7eiVk1r85hlio1_r1_500.gif');
  background-repeat: no-repeat;
  box-shadow: inset 0 20px 20px -20px $gif-inset,
              inset 20px 0 20px -20px $gif-inset,
              inset -20px 0 20px -20px $gif-inset,
              inset 0 -20px 20px -20px $gif-inset,
              0 0 10px 10px #d6c8afb6;
  height: 270px;
  width: 500px;
  margin: auto 25%;
}

#loading h1 {
  color: rgb(237,229,203);
  display: block;
  line-height: 135px;
  margin-left: 130px;
  margin-top: -38px;
  animation-duration: 3s;
  animation-name: slidein;
}

#loading h1 img {
  display: block;
  margin-left: 90px;  
  height: 90px;
  width: 90px;
}

@keyframes slidein {
  0% {
    margin-top: -700px;      
    margin-left: 130px;
  }

  100% {
    margin-left: 130px;
  }
}