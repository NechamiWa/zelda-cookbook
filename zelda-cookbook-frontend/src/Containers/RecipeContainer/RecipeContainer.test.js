/* eslint-disable */
import React from 'react';
import { shallow } from 'enzyme';
import { RecipeContainer, MSTP } from './RecipeContainer';

describe('RecipeContainer', () => {
  let wrapper;
  let mockIngredientFilter = []
  let mockNameFilter = ''
  let mockTypeFilter = ''
  let mockRecipes = [
    {
      id: 1,
      name: 'Baked Apple',
      type: 'none',
      ingredient1: 'Apple',
      ingredient2: null,
      ingredient3: null,
      ingredient4: null,
      ingredient5: null,
      ingredient1_id: 2,
      ingredient2_id: null,
      ingredient3_id: null,
      ingredient4_id: null,
      ingredient5_id: null,
    },
    {
      id: 2, 
      name: 'Apple Pie',
      ingredient1: 'Apple',
      type: 'Restore Hearts',
      ingredient2: 'Tabantha Wheat',
      ingredient3: 'Cane Sugar',
      ingredient4: 'Goat Butter',
      ingredient5: null,
      ingredient1_id: 2,
      ingredient2_id: 3,
      ingredient3_id: 4,
      ingredient4_id: 5,
      ingredient5_id: null,
    }
  ]

  beforeEach(() => {
    wrapper = shallow(<RecipeContainer
      ingredientFilter={mockIngredientFilter}
      nameFilter={mockNameFilter}
      typeFilter={mockTypeFilter}
      recipes={mockRecipes}
    />);
  });

  it('should have a default state', () => {
    expect(wrapper.state().loading).toEqual(true)
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('filterByIngredient', () => {
    it('should return all recipes if there are no ingredients in the filter', () => {
      expect(
        wrapper
        .instance()
        .filterByIngredient(mockRecipes)
      ).toEqual(mockRecipes)
    })

    it('should return an array of recipes whose ingredients match what is in the filter', () => {
      const wrapper = shallow(<RecipeContainer
        ingredientFilter={[{name: 'Tabantha Wheat', id: '3'}]}
        nameFilter={mockNameFilter}
        typeFilter={mockTypeFilter}
        recipes={mockRecipes}
      />);
      
      expect(
        wrapper
        .instance()
        .filterByIngredient(mockRecipes)
      ).toEqual([{
        id: 2,
        name: 'Apple Pie',
        type: 'Restore Hearts',
        ingredient1: 'Apple',
        ingredient2: 'Tabantha Wheat',
        ingredient3: 'Cane Sugar',
        ingredient4: 'Goat Butter',
        ingredient5: null,
        ingredient1_id: 2,
        ingredient2_id: 3,
        ingredient3_id: 4,
        ingredient4_id: 5,
        ingredient5_id: null,
      }])
    })
  })

  describe('filterByName', () => {
    it('should return all ingredients if there is no name in filter', () => {
      expect(
        wrapper
        .instance()
        .filterByName(mockRecipes)
      ).toEqual(mockRecipes)
    })

    it('should return an array with ingredients matching the name filter', () => {
      const wrapper = shallow(<RecipeContainer
        ingredientFilter={mockIngredientFilter}
        nameFilter={'Pie'}
        typeFilter={mockTypeFilter}
        recipes={mockRecipes}
      />);

      expect(
        wrapper
        .instance()
        .filterByName(mockRecipes)
      ).toEqual([{
        id: 2,
        name: 'Apple Pie',
        type: 'Restore Hearts',
        ingredient1: 'Apple',
        ingredient2: 'Tabantha Wheat',
        ingredient3: 'Cane Sugar',
        ingredient4: 'Goat Butter',
        ingredient5: null,
        ingredient1_id: 2,
        ingredient2_id: 3,
        ingredient3_id: 4,
        ingredient4_id: 5,
        ingredient5_id: null,
      }])
    })

    it('should not be case sensative', () => {
      const wrapper = shallow(<RecipeContainer
        ingredientFilter={mockIngredientFilter}
        nameFilter={'pIE'}
        typeFilter={mockTypeFilter}
        recipes={mockRecipes}
      />);

      expect(
        wrapper
        .instance()
        .filterByName(mockRecipes)
      ).toEqual([{
        id: 2,
        name: 'Apple Pie',
        type: 'Restore Hearts',
        ingredient1: 'Apple',
        ingredient2: 'Tabantha Wheat',
        ingredient3: 'Cane Sugar',
        ingredient4: 'Goat Butter',
        ingredient5: null,
        ingredient1_id: 2,
        ingredient2_id: 3,
        ingredient3_id: 4,
        ingredient4_id: 5,
        ingredient5_id: null,
      }])
    })
  })

  describe('filterByType', () => {
    it('should return all ingredients if there is no filter applied', () => {
      expect(wrapper.instance().filterByType(mockRecipes)).toEqual(mockRecipes);
    })

    it('should return an array of ingredients that match the type filter', () => {
      const wrapper = shallow(<RecipeContainer
        ingredientFilter={mockIngredientFilter}
        nameFilter={mockNameFilter}
        typeFilter={'Restore Hearts'}
        recipes={mockRecipes}
      />);

      expect(wrapper.instance().filterByType(mockRecipes)).toEqual([{
        id: 2,
        name: 'Apple Pie',
        type: 'Restore Hearts',
        ingredient1: 'Apple',
        ingredient2: 'Tabantha Wheat',
        ingredient3: 'Cane Sugar',
        ingredient4: 'Goat Butter',
        ingredient5: null,
        ingredient1_id: 2,
        ingredient2_id: 3,
        ingredient3_id: 4,
        ingredient4_id: 5,
        ingredient5_id: null,
      }])
    })
  })

  describe('filterRecipes', () => {
    it('should call each filtering algorithm with the array of recipes', () => {
      wrapper.instance().filterByIngredient = jest.fn()
        .mockImplementation(() => mockRecipes)
      wrapper.instance().filterByName = jest.fn()
        .mockImplementation(() => mockRecipes)
      wrapper.instance().filterByType = jest.fn()
        .mockImplementation(() => mockRecipes)
      
      wrapper.instance().filterRecipes(mockRecipes)

      expect(wrapper.instance().filterByIngredient).toHaveBeenCalledWith(mockRecipes)
      expect(wrapper.instance().filterByName).toHaveBeenCalledWith(mockRecipes)
      expect(wrapper.instance().filterByType).toHaveBeenCalledWith(mockRecipes)
    })

    it('should return the array of recipes after passing through the filters', () => {
      wrapper.instance().filterByIngredient = jest.fn()
        .mockImplementation(() => mockRecipes)
      wrapper.instance().filterByName = jest.fn()
        .mockImplementation(() => mockRecipes)
      wrapper.instance().filterByType = jest.fn()
        .mockImplementation(() => mockRecipes)

      expect(wrapper.instance().filterRecipes(mockRecipes)).toEqual(mockRecipes);
    })
  })

  describe('renderRecipes', () => {
    it('should match the snapshot if there are no recipes to render', () => {
      const wrapper = shallow(<RecipeContainer
        ingredientFilter={mockIngredientFilter}
        nameFilter={mockNameFilter}
        typeFilter={mockTypeFilter}
        recipes={[]}
      />);

      expect(wrapper.instance().renderRecipes()).toMatchSnapshot();
    })

    it('should match the snapshot when it renders ingredients', () => {
      const wrapper = shallow(<RecipeContainer
        ingredientFilter={mockIngredientFilter}
        nameFilter={mockNameFilter}
        typeFilter={mockTypeFilter}
        recipes={mockRecipes}
      />);

      expect(wrapper.instance().renderRecipes()).toMatchSnapshot()
    })
  })

  describe('MSTP', () => {
    it('should return an object with props from state', () => {
      const mockState = {
        recipes: [],
        ingredientFilter: [],
        nameFilter: '',
        typeFilter: ''
      }

      const mapped = MSTP(mockState);

      expect(mapped.recipes).toEqual([]);
      expect(mapped.ingredientFilter).toEqual([]);
      expect(mapped.nameFilter).toEqual('');
      expect(mapped.typeFilter).toEqual('');
    })
  })
});