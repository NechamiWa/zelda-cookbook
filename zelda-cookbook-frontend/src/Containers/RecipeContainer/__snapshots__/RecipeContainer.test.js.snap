// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeContainer renderRecipes should match the snapshot if there are no recipes to render 1`] = `
<div
  id="loading"
>
  <h1>
    No Recipe Found!
    <img
      alt="Dubious Food"
      src="DubiousFood.png"
    />
  </h1>
</div>
`;

exports[`RecipeContainer renderRecipes should match the snapshot when it renders ingredients 1`] = `
Array [
  <RecipeCard
    id={1}
    ingredient1="Apple"
    ingredient1_id={2}
    ingredient2={null}
    ingredient2_id={null}
    ingredient3={null}
    ingredient3_id={null}
    ingredient4={null}
    ingredient4_id={null}
    ingredient5={null}
    ingredient5_id={null}
    name="Baked Apple"
    type="none"
/>,
  <RecipeCard
    id={2}
    ingredient1="Apple"
    ingredient1_id={2}
    ingredient2="Tabantha Wheat"
    ingredient2_id={3}
    ingredient3="Cane Sugar"
    ingredient3_id={4}
    ingredient4="Goat Butter"
    ingredient4_id={5}
    ingredient5={null}
    ingredient5_id={null}
    name="Apple Pie"
    type="Restore Hearts"
/>,
]
`;

exports[`RecipeContainer should match the snapshot 1`] = `
<div
  id="all-recipe-container"
>
  <RecipeCard
    id={1}
    ingredient1="Apple"
    ingredient1_id={2}
    ingredient2={null}
    ingredient2_id={null}
    ingredient3={null}
    ingredient3_id={null}
    ingredient4={null}
    ingredient4_id={null}
    ingredient5={null}
    ingredient5_id={null}
    key="1"
    name="Baked Apple"
    type="none"
  />
  <RecipeCard
    id={2}
    ingredient1="Apple"
    ingredient1_id={2}
    ingredient2="Tabantha Wheat"
    ingredient2_id={3}
    ingredient3="Cane Sugar"
    ingredient3_id={4}
    ingredient4="Goat Butter"
    ingredient4_id={5}
    ingredient5={null}
    ingredient5_id={null}
    key="2"
    name="Apple Pie"
    type="Restore Hearts"
  />
</div>
`;
